fdbserver/storageserver.actor.cpp

```diff
@@ -4906,6 +4981,523 @@ ACTOR Future<Void> validateRangeAgainstServers(StorageServer* data,
+			TraceEvent(SevVerbose, "AuditStorageGetThisServerKeysFromKeyServersReadRangePartial", thisServerID)
+			    .detail("Range", range)
+			    .detail("PartialRange", partialCompleteRange)
+			    .detail("Prefix", serverKeysPrefixFor(thisServerID))
+			    .detail("ResultSize", readResult.size())
+			    .detail("AduitServerID", thisServerID);
+		}
+
+		Future<Version> getGrvF = tr->getReadVersion();
+		ASSERT(getGrvF.isReady());
+		readAtVersion = getGrvF.get();
+		res = std::make_tuple(partialCompleteRange, readAtVersion, ownRanges);
+		TraceEvent(SevInfo, "AuditStorageGetThisServerKeysFromKeyServersEnd", thisServerID)
+		    .detail("Range", range)
+		    .detail("AtVersion", readAtVersion)
+		    .detail("OwnRanges", ownRanges.size())
+		    .detail("ShardsInAnonymousPhysicalShardCount", shardsInAnonymousPhysicalShardCount)
+		    .detail("TotalShardsCount", totalShardsCount);
+
+	} catch (Error& e) {
+		TraceEvent(SevDebug, "AuditStorageGetThisServerKeysFromKeyServersError", thisServerID)
+		    .errorUnsuppressed(e)
+		    .detail("AuditServer", thisServerID.first());
+		throw e;
+	}
+
+	return res;
+}
+
+ACTOR Future<Void> auditStorageStorageServerShardQ(StorageServer* data, AuditStorageRequest req) {
+	state UID thisServerID = data->thisServerID;
+	TraceEvent(SevInfo, "AuditStorageShardStorageServerShardBegin", thisServerID);
+	ASSERT(req.getType() == AuditType::ValidateStorageServerShard);
+	wait(data->serveAuditStorageParallelismLock.take(TaskPriority::DefaultYield));
+	state FlowLock::Releaser holder(data->serveAuditStorageParallelismLock);
+	state Version versionBeforeWait;
+	state std::vector<std::pair<Version, KeyRangeRef>> shardAssignments;
+	state AuditStorageState res(req.id, thisServerID, req.getType());
+	state std::vector<std::string> errors;
+	state Transaction tr(data->cx);
+	state std::tuple<Optional<KeyRange>, Version, std::vector<KeyRange>> serverKeyRes;
+	state Version serverKeyReadAtVersion;
+	state Optional<KeyRange> serverKeyPartialReadRange;
+	// We want to find out any mismatch between ownRangesSeenByServerKey and ownRangesLocalView
+	state std::vector<KeyRange> ownRangesSeenByServerKey;
+	state std::vector<KeyRange> ownRangesLocalView;
+	state std::string failureReason;
+	state KeyRange shardInfoReadRange;
+
+	try {
+		if (data->version.get() == 0) {
+			failureReason = "SS version is 0";
+			throw audit_storage_failed();
+		}
+
+		// Read serverKeys and shardInfo
+		tr.setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+		wait(store(serverKeyRes, getThisServerKeysFromServerKeys(data, &tr, req.range)));
+		serverKeyPartialReadRange = std::get<0>(serverKeyRes);
+		serverKeyReadAtVersion = std::get<1>(serverKeyRes);
+		ownRangesSeenByServerKey = std::get<2>(serverKeyRes);
+		if (!ownRangesSeenByServerKey.empty()) {
+			// We want transaction read always succeeds
+			// Thus, any fall behind SS can eventually read the local data by retries
+			if (serverKeyReadAtVersion < data->version.get()) {
```

Comment:
`serverKeyReadAtVersion < data->version.get()` is likely to happen, since SS eagerly fetches mutations from TLogs, shall we retry `reading from serverKeys` while keeping the local info:

```
state Version localMinVersion = data->version.get();
while (serverKeyReadAtVersion < localMinVersion) {
  wait(delay(`sometime`));
  getThisServerKeysFromServerKeys()
}

Then we have `serverKeyReadAtVersion >= localMinVersion`, which can be merged to the code below.

Reply:
YEs. This looks better